Store ClipScrollNodes in a vector

mrobinson opened the pull request:
Instead of storing ClipScrollNodes in a HashMap, store them in a vector.
This removes the CPU cost of accessing any particular ClipScrollNode. In
its place is a somewhat clunky system of translating a ClipId into a
ClipScrollIndex. These are allocated based on the total number of
ClipIds per display list, so the translation should be fairly cheap. A
per-pipeline offset and cache is stored, so the number of HashMap
lookups for display list flattening should be the same order as the
number of pipelines.

The big trade off here is that when descending into iframes, we need to
allocate space in the ClipScrollNode array ahead of time. If these
memory allocations turn out to be a problem, we can modify the code to
try to mutate existing nodes or to use a more complicated system of sub
vectors or slices.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/2449)
<!-- Reviewable:end -->

mrobinson commented:
Sorry for the size of this PR. It's difficult to make such a deep change without changing a lot of code.

Here is the Gecko try job for this change: https://treeherder.mozilla.org/#/jobs?repo=try&revision=8950bf1f433b3fd3b71b5cb526d9cb71fea98770&selectedJob=163434876

mrobinson commented:
@staktrace Heads up about the Gecko changes. I had to modify gfx/webrender_bindings/webrender_ffi_generated.h manually because it seems that cbindgen has trouble with the current WebRender source tree.

staktrace commented:
I feel like the gecko-side changes should either assert that u64 and usize have the same number of bits (which might not actually be true on all platforms) or better, we should replace uint64_t with size_t wherever we're keeping WR clipids in C++. This is probably hard to do at the moment but will be a lot simpler if I land my changes for using ClipChains (but that's blocked on #2388).

kvark commented:
Looks like a decent optimization, at the cost of increased re-direction complexity of the code.

---

Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[webrender/src/clip_scroll_tree.rs, line 135 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tunaHgaXZyMB0lHlf:-L5tunaHgaXZyMB0lHlg:b9mr4ub) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/clip_scroll_tree.rs#L135)):*
> ```Rust
>         // TODO(mrobinson): We should eventually make this impossible to misuse.
>         debug_assert!(self.nodes.len() >= 1);
>         ClipScrollNodeIndex(1)
> ```

didn't we have constants defined for those special indices somewhere?

---

*[webrender/src/clip_scroll_tree.rs, line 466 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tvf6oX2IXZeV84hl8:-L5tvf6oX2IXZeV84hl9:b-vuqirc) ([raw file](https://github.com/servo/webrender/blob/c7778751c13ebe40c562e174c19a0e95cdc6a895/webrender/src/clip_scroll_tree.rs#L466)):*
> ```Rust
> 
> 
>             if let Some((offset, clamping)) = self.pending_scroll_offsets.remove(&external_id) {
> ```

why is this removed?

---

*[webrender/src/clip_scroll_tree.rs, line 505 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tw8rNtU1SD_gB4M9V:-L5tw8rNtU1SD_gB4M9W:b-qzxvem) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/clip_scroll_tree.rs#L505)):*
> ```Rust
>         }
> 
>         if index.0 == self.nodes.len() {
> ```

I'd refactor that code into:
```rust
if let Some(ref mut empty_node) = self.nodes.get_mut(index.0) {
    *empty_node = node;
    return
}
self.nodes.reserve_exact(index.0 + 1 - self.nodes.len());
self.nodes.extend((self.nodes.len() .. index.0).map(|_| ClipScrollNode::empty());
self.nodes.push(node);
```

---

*[webrender/src/clip_scroll_tree.rs, line 511 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tw1_J8sHY5XFZ3rYy:-L5tw1_J8sHY5XFZ3rYz:bhli9mc) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/clip_scroll_tree.rs#L511)):*
> ```Rust
> 
>         if index.0 > self.nodes.len() {
>             self.nodes.reserve_exact(index.0);
> ```

this is incorrect usage of `reserve_exact`. The parameter is meant to be the additional count, not total count

---

*[webrender/src/frame.rs, line 49 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tx-UTOIkaqqtxAU_d:-L5tx-UTOIkaqqtxAU_e:b1lomuw) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L49)):*
> ```Rust
>     cached_index: Option<(ClipId, ClipChainIndex)>,
>     pipeline_offsets: FastHashMap<PipelineId, usize>,
>     cached_pipeline_offset: Option<(PipelineId, usize)>,
> ```

need to comment cleanly the difference between `cached_pipeline_offset` and `pipeline_offsets`

---

*[webrender/src/frame.rs, line 54 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tx49uxQjmDl-2peTb:-L5tx49uxQjmDl-2peTc:b-k2m4mu) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L54)):*
> ```Rust
> 
> impl ClipIdToIndexMapper {
>     fn new() -> ClipIdToIndexMapper {
> ```

nit: `Self` ftw!

---

*[webrender/src/frame.rs, line 58 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tx6zka9GmAollWCDz:-L5tx6zka9GmAollWCE-:b-w87wzn) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L58)):*
> ```Rust
>             clip_chain_map: FastHashMap::default(),
>             cached_index: None,
>             pipeline_offsets: FastHashMap::default(),
> ```

the whole implementation can probably be replaced by `derive(Default)`

---

*[webrender/src/frame.rs, line 60 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tyARsmf6OFlkRhP0T:-L5tyARsmf6OFlkRhP0U:b6plx5j) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L60)):*
> ```Rust
>             pipeline_offsets: FastHashMap::default(),
>             cached_pipeline_offset: None,
>             next_available_offset: 0,
> ```

what's the lifetime of those fields? I don't see them being reset

---

*[webrender/src/frame.rs, line 83 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5txIL3YKEiHD2OSWou:-L5txIL3YKEiHD2OSWov:b-pu1kbe) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L83)):*
> ```Rust
>     }
> 
>     pub fn get_clip_chain_index_and_cache_result(&mut self, id: &ClipId) -> ClipChainIndex {
> ```

nit: maybe `cache_clip_chain_index`? to indicate that it's mutating

---

*[webrender/src/frame.rs, line 109 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5ty1JmDBnnxxEum6KJ:-L5ty1JmDBnnxxEum6KK:bk0fci9) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L109)):*
> ```Rust
>         let (index, pipeline_id) = match id {
>             ClipId::Clip(index, pipeline_id) => (index, pipeline_id),
>             ClipId::ClipChain(_) => unreachable!("Tried to use ClipChain as scroll node."),
> ```

not sure if this is truly `unreachable`. Sounds more like a `panic` to me

---

*[webrender/src/hit_test.rs, line 130 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tzYso8UnEujtYlaCJ:-L5tzYso8UnEujtYlaCK:b-ojrgmj) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/hit_test.rs#L130)):*
> ```Rust
>             let index = ClipScrollNodeIndex(index);
>             if !self.pipeline_root_nodes.contains_key(&node.pipeline_id) {
>                 self.pipeline_root_nodes.insert(node.pipeline_id, index);
> ```

nit: use `Entry` API

---

*[webrender_api/src/display_item.rs, line 771 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tzmhF-REMRcTFeMXN:-L5tzmhF-REMRcTFeMXO:bccd4je) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender_api/src/display_item.rs#L771)):*
> ```Rust
> 
> #[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
> pub enum ClipId {
> ```

This is significantly less used now. We should take another look at the type and see if it's still needed, or maybe rename it to something more verbose

---

*[webrender_api/src/display_list.rs, line 518 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5u-3KkpU-MM65Lo9N8:-L5u-3KkpU-MM65Lo9N9:bcn3ozs) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender_api/src/display_list.rs#L518)):*
> ```Rust
>                     },
>                     ClipChain(specific_item, clip_chain_ids) => {
>                         DisplayListBuilder::push_iter_impl(&mut temp, clip_chain_ids);
> ```

clip chain doesn't count for total clips?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2449#-:-L5u-CzZ9dHXuEthMbM6:bvxuvvj)*
<!-- Sent from Reviewable.io -->

mrobinson commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[webrender/src/clip_scroll_tree.rs, line 135 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tunaHgaXZyMB0lHlf:-L5weAfmJE0Gwak_GejB:bxschcs) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/clip_scroll_tree.rs#L135)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

didn't we have constants defined for those special indices somewhere?
</blockquote></details>

We have constants for ClipId, but they are semantically different. This is a good idea though, so I've added constants for these ones in `clip_scroll_tree.rs`.

---

*[webrender/src/clip_scroll_tree.rs, line 466 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tvf6oX2IXZeV84hl8:-L5weIqk0inE_XhkYQFf:b-93sh2n) ([raw file](https://github.com/servo/webrender/blob/c7778751c13ebe40c562e174c19a0e95cdc6a895/webrender/src/clip_scroll_tree.rs#L466)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

why is this removed?
</blockquote></details>

I had fixed this, but the fix was sitting uncommitted in my working tree. ☹

---

*[webrender/src/clip_scroll_tree.rs, line 505 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tw8rNtU1SD_gB4M9V:-L5weONjNSCmaBFdS5tU:bpzr73x) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/clip_scroll_tree.rs#L505)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

I'd refactor that code into:
```rust
if let Some(ref mut empty_node) = self.nodes.get_mut(index.0) {
    *empty_node = node;
    return
}
self.nodes.reserve_exact(index.0 + 1 - self.nodes.len());
self.nodes.extend((self.nodes.len() .. index.0).map(|_| ClipScrollNode::empty());
self.nodes.push(node);
```
</blockquote></details>

Okay. This seems fine to me as well. I had to make a couple minor changes to make the borrow checker happy. I've also added a comment here explaining why we have to do this instead of `Vec::resize`.

---

*[webrender/src/clip_scroll_tree.rs, line 511 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tw1_J8sHY5XFZ3rYy:-L5weZcAGcc7bdra-313:b-73m6vj) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/clip_scroll_tree.rs#L511)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

this is incorrect usage of `reserve_exact`. The parameter is meant to be the additional count, not total count
</blockquote></details>

Nice catch.

---

*[webrender/src/frame.rs, line 49 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tx-UTOIkaqqtxAU_d:-L5wea85BUvcnKz5dZ3J:b-5e5gwv) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L49)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

need to comment cleanly the difference between `cached_pipeline_offset` and `pipeline_offsets`
</blockquote></details>

I added comments for all the members and renamed `cached_index` to `cached_clip_chain_index`. 

---

*[webrender/src/frame.rs, line 54 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tx49uxQjmDl-2peTb:-L5wehi_KtqbHrKv3g71:b-896fix) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L54)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

nit: `Self` ftw!
</blockquote></details>

Done.

---

*[webrender/src/frame.rs, line 58 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tx6zka9GmAollWCDz:-L5wfHVdBzKOjYkjDslH:b-896fix) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L58)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

the whole implementation can probably be replaced by `derive(Default)`
</blockquote></details>

Done.

---

*[webrender/src/frame.rs, line 60 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tyARsmf6OFlkRhP0T:-L5wfK8e6RRbRGnseBZ0:budhoc6) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L60)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

what's the lifetime of those fields? I don't see them being reset
</blockquote></details>

This lives as long as the FlattenContext which is only alive as long as display list flattening is happening. This struct is used only be used once and then discarded.

---

*[webrender/src/frame.rs, line 83 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5txIL3YKEiHD2OSWou:-L5wffOD3j9SQiudxcC1:bm16o2c) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L83)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

nit: maybe `cache_clip_chain_index`? to indicate that it's mutating
</blockquote></details>

Hrm. I'm not sure because this is doing two things getting the index and caching it. I'm not sure if `get_and_cache_clip_chain_index` is any clearer, but maybe that's another option.

---

*[webrender/src/frame.rs, line 109 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5ty1JmDBnnxxEum6KJ:-L5wfr72oHvc05OkHIHW:b-896fix) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/frame.rs#L109)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

not sure if this is truly `unreachable`. Sounds more like a `panic` to me
</blockquote></details>

Done.

---

*[webrender/src/hit_test.rs, line 130 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tzYso8UnEujtYlaCJ:-L5wfsNnn36Tev0irZmM:b-896fix) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender/src/hit_test.rs#L130)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

nit: use `Entry` API
</blockquote></details>

Done.

---

*[webrender_api/src/display_item.rs, line 771 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5tzmhF-REMRcTFeMXN:-L5wh_T0S2AEHCx-jm1P:b-tbzpol) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender_api/src/display_item.rs#L771)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

This is significantly less used now. We should take another look at the type and see if it's still needed, or maybe rename it to something more verbose
</blockquote></details>

I confirmed that these are all still necessary.

---

*[webrender_api/src/display_list.rs, line 518 at r1](https://reviewable.io:443/reviews/servo/webrender/2449#-L5u-3KkpU-MM65Lo9N8:-L5whceR1c6kOEwZiqWJ:bpdrkrk) ([raw file](https://github.com/servo/webrender/blob/4df8a29aa9c14595cc105652d85138f659dd54ce/webrender_api/src/display_list.rs#L518)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

clip chain doesn't count for total clips?
</blockquote></details>

ClipChains don't have corresponding ClipScrollNodes so they don't increase the count.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2449)*
<!-- Sent from Reviewable.io -->

mrobinson commented:
@staktrace I modified the Gecko-side patch to store size_t for ids that correspond to ClipScrollNodes. It wasn't a lot of work (though did require me to edit more of the auto-generated bindings by hand).

The Gecko try for this modified version is here: https://treeherder.mozilla.org/#/jobs?repo=try&revision=acdbaae3df471bd15eb355520144af790a5e9b13

I still plan to look at #2388 today, but maybe it doesn't need to block this change?

staktrace commented:
Looks like your try push has compilation failures so the patch might need some tweaking. Also FYI we have a new cbindgen version posted which should allow you to regenerate the bindings without manual tinkering. You'll need to install/run it with rust nightly though:
```
cargo +nightly install --force cbindgen
cd $ROOT_OF_MOZILLA_TREE
rustup run nightly cbindgen toolkit/library/rust/ --crate webrender_bindings -o gfx/webrender_bindings/webrender_ffi_generated.h
```

Agreed that #2388 doesn't need to block this - but I'd like to review the final gecko-side changes for this PR once you have the compilation failures sorted out. Assuming it's a straightforward replacement of uint64_t to size_t in the relevant places it should be fine.

mrobinson commented:
@staktrace Thanks for the information about cbindgen. I've kicked off another Gecko try that should hopefully fix the build failures on 32-bit Linux: https://treeherder.mozilla.org/#/jobs?repo=try&revision=cd9b9ede957425da328fe7dd38fc215dc7475fbd

mrobinson commented:
I also confirmed that cbindgen didn't materially change code related to my change other than correcting a comment and fixing the implementation of the == operator override for BuiltDisplayList.

kvark commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2449#-:-L5ykBbU2Q7g6vahY6XD:bnfp4nl)*
<!-- Sent from Reviewable.io -->

kvark commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit b76ae01 has been approved by `kvark`

<!-- @bors-servo r=kvark b76ae0133211c708277671497d41421d1baaf548 -->

bors-servo commented:
:lock: Merge conflict

bors-servo commented:
:umbrella: The latest upstream changes (presumably #2443) made this pull request unmergeable. Please resolve the merge conflicts.

mrobinson commented:
@kvark @staktrace Thanks for looking at this PR. I've rebased and will land it now.

mrobinson commented:
@bors-servo r=kvark

bors-servo commented:
:pushpin: Commit b548082 has been approved by `kvark`

<!-- @bors-servo r=kvark b548082858dfd7a9be2b84c3dd6a457cf995ea75 -->

bors-servo commented:
:hourglass: Testing commit b548082858dfd7a9be2b84c3dd6a457cf995ea75 with merge 8e2672e52a5a80eba40030601653fe0c1ae223a1...

bors-servo commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/servo/webrender/build/1.0.8086), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/YO421I9ESX-cmTu4n9REBA), [status-travis](https://travis-ci.org/servo/webrender/builds/345200954?utm_source=github_status&utm_medium=notification)
Approved by: kvark
Pushing 8e2672e52a5a80eba40030601653fe0c1ae223a1 to master...
